### PR TITLE
Interval Building Phase 1

### PR DESCRIPTION
**Pull request by dgoodwin**
[TRT-1147](https://issues.redhat.com//browse/TRT-1147)

Intervals are an important part of our debugging toolkit, but suffer from some shortcomings in their design that limit their use. (examples: e2e tests with spaces and quotations in locators making them hard/impossible to parse, inability to carry request audit IDs with messages without breaking disruption checks to see if the error changed, etc)

Old interval:
```json
        {
            "level": "Info",
            "locator": "ns/openshift-e2e-loki pod/event-exporter-55f76dcbf6-rgvxz uid/d8508825-aa23-4549-b1cc-a9f737cec5ba",
            "message": "constructed/true reason/Created",
            "from": "2023-07-15T09:46:05Z",
            "to": "2023-07-15T10:10:31Z"
        },
```

Interval in this PR:

```json
        {
            "level": "Error",
            "locator": "node/ci-op-yv5xrq3b-206af-tjlcx-master-0",
            "message": "reason/FailedToAuthenticateWithOpenShiftUser Jul 21 17:53:17.491460 ci-op-yv5xrq3b-206af-tjlcx-master-0 kubenswrapper[2187]: E0721 17:53:17.374007    2187 event.go:280] Server rejected event '\u0026v1.Event{TypeMeta:v1.TypeMeta{Kind:\"\", APIVersion:\"\"}, ObjectMeta:v1.ObjectMeta{Name:\"ci-op-yv5xrq3b-206af-tjlcx-master-0.1773f37b98a2b824\", GenerateName:\"\", Namespace:\"default\", SelfLink:\"\", UID:\"\", ResourceVersion:\"\", Generation:0, CreationTimestamp:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), DeletionTimestamp:\u003cnil\u003e, DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string(nil), Annotations:map[string]string(nil), OwnerReferences:[]v1.OwnerReference(nil), Finalizers:[]string(nil), ManagedFields:[]v1.ManagedFieldsEntry(nil)}, InvolvedObject:v1.ObjectReference{Kind:\"Node\", Namespace:\"\", Name:\"ci-op-yv5xrq3b-206af-tjlcx-master-0\", UID:\"ci-op-yv5xrq3b-206af-tjlcx-master-0\", APIVersion:\"\", ResourceVersion:\"\", FieldPath:\"\"}, Reason:\"Starting\", Message:\"Starting kubelet.\", Source:v1.EventSource{Component:\"kubelet\", Host:\"ci-op-yv5xrq3b-206af-tjlcx-master-0\"}, FirstTimestamp:time.Date(2023, time.July, 21, 17, 53, 17, 361395748, time.Local), LastTimestamp:time.Date(2023, time.July, 21, 17, 53, 17, 361395748, time.Local), Count:1, Type:\"Normal\", EventTime:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), Series:(*v1.EventSeries)(nil), Action:\"\", Related:(*v1.ObjectReference)(nil), ReportingController:\"\", ReportingInstance:\"\"}': 'events is forbidden: User \"system:anonymous\" cannot create resource \"events\" in API group \"\" in the namespace \"default\"' (will not retry!)",
            "tempStructuredLocator": {
                "type": "Node",
                "keys": {
                    "node": "ci-op-yv5xrq3b-206af-tjlcx-master-0"
                }
            },
            "tempStructuredMessage": {
                "reason": "FailedToAuthenticateWithOpenShiftUser",
                "cause": "",
                "humanMessage": "Jul 21 17:53:17.491460 ci-op-yv5xrq3b-206af-tjlcx-master-0 kubenswrapper[2187]: E0721 17:53:17.374007    2187 event.go:280] Server rejected event '\u0026v1.Event{TypeMeta:v1.TypeMeta{Kind:\"\", APIVersion:\"\"}, ObjectMeta:v1.ObjectMeta{Name:\"ci-op-yv5xrq3b-206af-tjlcx-master-0.1773f37b98a2b824\", GenerateName:\"\", Namespace:\"default\", SelfLink:\"\", UID:\"\", ResourceVersion:\"\", Generation:0, CreationTimestamp:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), DeletionTimestamp:\u003cnil\u003e, DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string(nil), Annotations:map[string]string(nil), OwnerReferences:[]v1.OwnerReference(nil), Finalizers:[]string(nil), ManagedFields:[]v1.ManagedFieldsEntry(nil)}, InvolvedObject:v1.ObjectReference{Kind:\"Node\", Namespace:\"\", Name:\"ci-op-yv5xrq3b-206af-tjlcx-master-0\", UID:\"ci-op-yv5xrq3b-206af-tjlcx-master-0\", APIVersion:\"\", ResourceVersion:\"\", FieldPath:\"\"}, Reason:\"Starting\", Message:\"Starting kubelet.\", Source:v1.EventSource{Component:\"kubelet\", Host:\"ci-op-yv5xrq3b-206af-tjlcx-master-0\"}, FirstTimestamp:time.Date(2023, time.July, 21, 17, 53, 17, 361395748, time.Local), LastTimestamp:time.Date(2023, time.July, 21, 17, 53, 17, 361395748, time.Local), Count:1, Type:\"Normal\", EventTime:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), Series:(*v1.EventSeries)(nil), Action:\"\", Related:(*v1.ObjectReference)(nil), ReportingController:\"\", ReportingInstance:\"\"}': 'events is forbidden: User \"system:anonymous\" cannot create resource \"events\" in API group \"\" in the namespace \"default\"' (will not retry!)",
                "annotations": {
                    "reason": "FailedToAuthenticateWithOpenShiftUser"
                }
            },
            "from": "2023-07-21T17:53:17Z",
            "to": "2023-07-21T17:53:18Z"
        },
```

Goal in this PR is to move slowly to structured locators and messages both. This PR introduces the builders for condition/message/locator, and migrates SOME but not all uses of Conditions. Serialization temporarily includes structuredLocator and structuredMessage, which we hope will move to locator/message in the future and take over their plain string counterparts.

### CODE REVIEW

File: pkg/monitor/monitorapi/types.go [hunk 1]
```diff
@@ -60,20 +60,135 @@ func EventLevelFromString(s string) (EventLevel, error) {
...
-	Message string
+	// TODO: Goal here is to drop Locator/Message, and rename the structured variants to Locator/Message
+	Locator           string
+	StructuredLocator Locator
+	Message           string
+	StructuredMessage Message
 }
 
-type EventInterval struct {
+type LocatorType string
```
Comment (deads2k) [inline pkg/monitor/monitorapi/types.go:72]: huh, you or me?  This is a good idea, though we should kill "other" at some future point.

File: pkg/monitor/monitorapi/types.go [hunk 2]
```diff
@@ -60,20 +60,135 @@ func EventLevelFromString(s string) (EventLevel, error) {
...
+
+const (
+	ConstructionOwnerNodeLifecycle = "node-lifecycle-constructor"
+	ConstructionOwnerPodLifecycle  = "pod-lifecycle-constructor"
+)
+
+type Message struct {
+	// TODO: reason/cause both fields and annotations...
+	Reason       IntervalReason
+	Cause        string
```
Comment (deads2k) [inline pkg/monitor/monitorapi/types.go:177]: this smells like me.  At some future date I think we shoudl remove.

File: pkg/monitor/monitorapi/types.go [hunk 3]
```diff
@@ -60,20 +60,135 @@ func EventLevelFromString(s string) (EventLevel, error) {
 }
 
 type Condition struct {
```
Comment (deads2k) [inline pkg/monitor/monitorapi/types.go:62]: At some future date I think we'll want `RelatedObjects []Locator` to allow linkage of some kind.

Comment (deads2k) [inline pkg/monitor/monitorapi/types.go:62]: At some future date, I think we should have rendering hints for the timeline charts.

**Comment by deads2k** [comment]
`{  Watchdog alert not found}` failing in https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/28069/pull-ci-openshift-origin-master-e2e-gcp-ovn-upgrade/1681726289221783552 might be related.

File: pkg/monitor/monitorapi/types.go (hunk 2 above)
Comment (dgoodwin) [reply pkg/monitor/monitorapi/types.go:177]: Yes I was wondering it iwas intentional to have reason/cause both as fields and in annotations map. I will remove the fields.

**Comment by dgoodwin** [comment]
> `{ Watchdog alert not found}` failing in https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/28069/pull-ci-openshift-origin-master-e2e-gcp-ovn-upgrade/1681726289221783552 might be related.

Thanks!

**Comment by deads2k** [comment]
Maps aren't rendering in the event.json.

Other than watchdog and rendering, any reason this is WIP instead of pinning progress?

**Comment by DennisPeriquet** [comment]
Looking at one of the [presubmit json files](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/28069/pull-ci-openshift-origin-master-e2e-aws-ovn-single-node-upgrade/1683459537903816704/artifacts/e2e-aws-ovn-single-node-upgrade/single-node-e2e-test/artifacts/junit/),

For every type = "Node", you have a locator of the form `node/xxxxx`.

```
$ cat  e2e-timelines_spyglass_20230724-143826.json|jq '.items[] | select(.tempStructuredLocator.type == "Node")| .locator'
"node/ip-10-0-170-75.us-west-2.compute.internal"
"node/ip-10-0-170-75.us-west-2.compute.internal"
"node/ip-10-0-170-75.us-west-2.compute.internal"
```

There are also some cases where node is of that form but the type is set to `""`:

```
$ tail -19 e2e-timelines_spyglass_20230724-143826.json
        {
            "level": "Warning",
            "locator": "node/ip-10-0-170-75.us-west-2.compute.internal",
            "message": "constructed/node-lifecycle-constructor reason/NotReady role/ state/NotReady never completed",
            "tempStructuredLocator": {
                "type": "",
                "keys": null
            },
            "tempStructuredMessage": {
                "reason": "",
                "cause": "",
                "humanMessage": "",
                "annotations": null
            },
            "from": "2023-07-24T15:53:29Z",
            "to": "2023-07-24T15:53:37Z"
        }
    ]
}
```

Should that last one of `"type": "Node",`?

**Pull request merged**